### DISCUSSION TITLE
Batched Requests?

### DISCUSSION THREAD

**Discussion by burkgar** [ecederstrand/exchangelib#867] [Q&A]
Does exchangelib batch requests? I am having trouble using it with O365 because I am getting throttled. If the requests are being batched then I know there is probably not a way to fix the problem on my end, and will need to ask for less throttling.

**Comment by ecederstrand** [comment]
That depends on what you mean by batching. In the default configuration, we only open one TCP connection to the server. There is no concurrency going on. Which kind of errors do you see?

**Reply by burkgar** [reply]
I am getting an error stating "server requests back off until xxx time". I made a multithreaded script so it's going through several mailboxes at once and reading newly modified items since the last time that inbox was passed over.

**Reply by ecederstrand** [reply]
A server is allowed to request back-off, even though your server is not rate-limited by throttling policies on the server. Usually you'll get other error messages in that case. The error you're seeing could as well be a server that's having a hard time and requesting all clients to back off. exchangelib can handle these errors gracefully. See https://ecederstrand.github.io/exchangelib/#fault-tolerance

**Reply by burkgar** [reply]
Thank you for the response. I actually did use the fault tolerance settings in this script and have the max wait time set to 3600 as shown in the example. The script works for a little while and then starts getting the back off messages - that's why I have assumed it's throttling (plus I have been talking to Microsoft Support and they mentioned they do throttle).

My hope was that there might be something I could do to reduce the number of requests so that the back off requests will stop.

The basic flow of the script is a loop that connects to the API with oAuth, opens a mailbox, pulls the items modified since the last pass and reads several of their properties (ex: sender, categories, folder, last modified time etc.), and moves on to the next mailbox. 

Due to the use case for this script I need to be able to go through around 200+ mailboxes within a 15 minute time frame.

**Reply by ecederstrand** [reply]
If you haven't already, you can limit the amount of data transferred by using `.only()` to limit the fields you are fetching. Otherwise, you may need to limit the concurrency of your threaded app.

200 mailboxes in 15 minutes should be doable on O365.
